Trap ERR on Upgrade Script.

orsenthil opened the pull request:
## High Level Description

Trap ERR on the Upgrade Script and show more information and why the upgrade failed

## Related Issues

  - [DCOS_OSS-839](https://jira.dcos.io/browse/DCOS_OSS-839) 

## Checklist for all PR's

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

amitaekbote commented:
From what I understood the issue is that pkgpanda sends its errors to stdout and not to stderr, we should be catching any exception/errors here - https://github.com/dcos/dcos/blob/master/pkgpanda/actions.py#L22 and sending them to stderr. 

We send all stdout to /dev/null https://github.com/dcos/dcos/blob/master/dcos_installer/upgrade.py#L88 

@branden what do you think?

branden commented:
@amitaekbote The upgrade script uses the pkgpanda installed on the cluster node, so fixing this in pkgpanda would require a new 1.8 release. IMO we should make pkgpanda print error output to stderr for upgrades going forward, but for the 1.8 -> 1.9 release this is likely the best we can do.

orsenthil commented:
@mesosphere-mergebot bump-ee

orsenthil commented:
This actually works. 

```
[centos@ip-10-10-0-221 ~]$ sudo bash ./dcos_node_upgrade.sh
Upgrading DC/OS master 1.9.0 -> 1.9-dev
Upgrade Script failed at line: 79
```

I think, I will desire more information from the upgrade failure. Let me see what further can be added.

orsenthil commented:
I like this output better.

```
$ sudo bash ./dcos_node_upgrade.sh
Upgrading DC/OS master 1.9.0 -> 1.9-dev
Script failed while executing:
pkgpanda activate --no-block dcos-metadata--setup_023a92510e0feefd9282108c6baf17b951602a86 python-markupsafe--dd46d2a3c58611656a235f96d4adc51b2a7a590e python-gunicorn--a537f95661fb2689c52fe12510eb0d01cb83af60 python-jinja2--7450f5ae5a822f63f7a58c717207be0456df51ed python-pyyaml--81dd44cc4a24db7cefa7016c6586a131acf279c3 python-isodate--40d378c688e6badfd16676dd8b51b742bfebc8d5 mesos--aaedd03eee0d57f5c0d49c74ff1e5721862cad98 dcos-image--eefcde959b01a15a31ce0a9b9813d9bac53b4f3a dcos-ui--d4afd695796404a5b35950c3daddcae322481ac4 dcos-config--setup_023a92510e0feefd9282108c6baf17b951602a86 dcos-signal--1bcd3b612cbdc379380dcba17cdf9a3b6652d9dc python-dateutil--fdc6ff929f65dd0918cf75a9ad56704683d31781 openssl--b01a32a42e3ccba52b417276e9509a441e1d4a82 cosmos--20decef90f0623ed253a12ec4cf5c148b18d8249 dnspython--0f833eb9a8abeba3179b43f3a200a8cd42d3795a dcos-image-deps--83584fd868e5b470f7cf754424a9a75b328e9b68 spartan--9cc57a3d55452b905d90e3201f56913140914ecc boto--6344d31eef082c7bd13259b17034ea7b5c34aedf bootstrap--95f349440418f91b1c329d48c1dc6afc9428316e check-time--be7d0ba757ec87f9965378fee7c76a6ee5ae996d strace--7d01796d64994451c1b2b82d161a335cbe90569b python-kazoo--cb7ce13a1068cd82dd84ea0de32b529a760a4bdd teamcity-messages--e623a4d86eb3a8d199cefcc240dd4c5460cb2962 java--cd5e921ce66b0d3303883c06d73a657314044304 mesos-modules--311849eaae42696b8a7eefe86b9ab3ebd9bd48f5 python-requests--1b2cadbd3811cc0c2ee235ce927e13ea1d6af41d flask--26d1bcdb2d1c3dcf1d2c03bc0d4f29c86d321b21 curl--fc3486c43f98e63f9b12675f1356e8fe842f26b0 toybox--f235594ab8ea9a2864ee72abe86723d76f92e848 pytest--78aee3e58a049cdab0d266af74f77d658b360b4f erlang--a9ee2530357a3301e53056b36a93420847b339a3 pkgpanda-role--f8a749a4a821476ad2ef7e9dd9d12b6a8c4643a4 boost-libs--2015ccb58fb756f61c02ee6aa05cc1e27459a9ec 3dt--7847ebb24bf6756c3103902971b34c3f09c3afbd cni--e48337da39a8cd379414acfe0da52a9226a10d24 dcos-metrics--23ee2f89c58b1258bc959f1d0dd7debcbb3d79d2 adminrouter--0493a6fdaed08e1971871818e194aa4607df4f09 ncurses--d889894b71aa1a5b311bafef0e85479025b4dacb six--f06424b68523c4dfa2a7c3e7475d479f3d361e42 python-tox--07244f8a939a10353634c952c6d88ec4a3c05736 octarine--e86d3312691b12523280d56f6260216729aaa0ad dvdcli--5374dd4ffb519f1dcefdec89b2247e3404f2e2e3 python-cryptography--4184767c68e48801dd394072cb370c610a05029d libsodium--9ff915db08c6bba7d6738af5084e782b13c84bf8 python-azure-mgmt-resource--03c05550f43b0e7a4455c33fe43b0deb755d87f0 navstar--1128db0234105a64fb4be52f4453cd6aa895ff30 dcos-history--3b0d303c523243560c6aad7a5f6667f001c2592c python-retrying--eb7b8bac133f50492b1e1349cbe77c3e38bd02c3 python-docopt--beba78faa13e5bf4c52393b4b82d81f3c391aa65 libffi--83ce3bd7eda2ef089e57efd2bc16c144d5a1f094 mesos-dns--0401501b2b5152d01bfa84ff6d007fdafe414b16 libevent--208be855d2be29c9271a7bd6c04723ff79946e02 dcos-log--4d630df863228f38c6333e44670b4c4b20a74832 rexray--869621bb411c9f2a793ea42cdfeed489e1972aaa avro-cpp--760c214063f6b038b522eaf4b768b905fed56ebc pkgpanda-api--92ecc32a52a1025cc8d1d58d9d57c3124836a8e1 python--b7a144a49577a223d37d447c568f51330ee95390 metronome--467e4c64f804dbd4cd8572516e111a3f9298c10d python-passlib--802ec3605c0b82428fedba60983b1bafaa036bb8 dcos-integration-test--c28bcb2347799dca43083f55e4c7b28503176f9c logrotate--7f7bc4416d3ad101d0c5218872858483b516be07 marathon--bfb24f7f90cb3cd52a1cb22a07caafa5013bba21 exhibitor--72d9d8f947e5411eda524d40dde1a58edeb158ed docker-gc--59a98ed6446a084bf74e4ff4b8e3479f59ea8528 dcos-oauth--0079529da183c0f23a06d2b069721b6fa6cc7b52 > /dev/null
```

Going to change to this and submit for review.

orsenthil commented on dcos_installer/upgrade.py:
```diff
@@ -12,14 +12,14 @@
 
-node_upgrade_template = """
-#!/bin/bash
+node_upgrade_template = """#!/bin/bash
```
This is required / better. SHEBANG line should be the first one.
In the current state.

```
$ head dcos_node_upgrade.sh

#!/bin/bash
#
# BASH script to upgrade DC/OS on a node
#
```

orsenthil commented on dcos_installer/upgrade.py:
```diff
@@ -38,6 +38,15 @@
+# bash should trap the ERR signal and run err_report, which prints the line of error
+# to STDERR.
+
+err_report() {
```
This trick is from here: http://stackoverflow.com/a/42219754/18852

Short snippets fall under fair use clause. (I can't think of a better way ignoring this).

orsenthil commented:
This is ready for review @branden , @amitaekbote , @lingmann .

branden commented:
The full `pkgpanda activate` line is a lot of text to ask a user to sift through in order to determine what happened, and unless they're a DC/OS provisioning expert, it's probably meaningless to them. I like the short error message with a line number. It's easier to tell at a glance that the script failed, and we can open the script to get the line that failed for further debugging.

I'd suggest something like this:

```
[centos@ip-10-10-0-221 ~]$ sudo bash ./dcos_node_upgrade.sh
Upgrading DC/OS master 1.9.0 -> 1.9-dev
ERROR: Upgrade script failed at line 79
```

lingmann commented:
I'm inclined to agree with @branden on printing a shorter error message with just the line number vs the code that failed. That would also removes the dependency on `sed` (which is notoriously bad from a cross-platform compatibility perspective) and simplify the echo statement. For example:
```
echo "ERROR: Upgrade failed at line ${BASH_LINENO}"
```

orsenthil commented:
Addressed review comments. New output.

```
$ sudo bash dcos_node_upgrade.sh
Upgrading DC/OS master 1.9.0 -> 1.9-dev
ERROR: Upgrade failed at line 78
```

branden commented on dcos_installer/upgrade.py:
```diff
@@ -38,6 +38,15 @@
+# of failure to STDERR.
+
+err_report() {
+        echo "ERROR: Upgrade failed at line ${BASH_LINENO}"
```
The rest of the script uses 4-space indentation so let's use that here as well.

orsenthil replied:
Addressed.

adam-mesos commented:
@lingmann Do we need to backport this to 1.9.1 too?

orsenthil commented:
@adam-mesos it's a good idea. I will create a backport PR.
